django/utils/translation/trans_real.py

```diff
@@ -465,11 +466,16 @@ def get_language_from_path(path, strict=False):
     """
     regex_match = language_code_prefix_re.match(path)
     if not regex_match:
+        if is_language_prefix_patterns_used(None, unprefixed=True):
```

Comment:
Hmm we can't pass `None` here, it has to be the request's `urlconf`.

I suggest we make `is_language_prefix_patterns_used` return a tuple of boolean as follow:

```python
def is_language_prefix_patterns_used(urlconf=None):
    for url_pattern in get_resolver(urlconf).url_patterns:
        if isinstance(url_pattern, LocaleRegexURLResolver):
            return True, url_pattern.prefix_default_language
        return False, False    
```

Completely remove these branch from `get_language_from_path()` and adjust `LocaleMiddleware` to correctly deal with `get_language_from_path` and `get_language_from_request` returning `None`:

```python
def process_request(self, request):
    urlconf = getattr(request, 'urlconf', settings.ROOT_URLCONF)
    i18n_patterns_used, prefixed_default_language = is_language_prefix_patterns_used(urlconf)
    language = translation.get_language_from_request(
        request, check_path=i18n_patterns_used
    )
    if not language and i18n_patterns_used and not prefixed_default_language:
        language = settings.DEFAULT_LANGUAGE
    translation.activate(language)
    request.LANGUAGE_CODE = translation.get_language()

def process_response(self, request, response):
    language = translation.get_language()
    language_from_path = translation.get_language_from_path(request.path_info)
    urlconf = getattr(request, 'urlconf', settings.ROOT_URLCONF)
    i18n_patterns_used, prefixed_default_language = is_language_prefix_patterns_used(urlconf)
    if not language_from_path and i18n_patterns_used and not prefixed_default_language:
        language_from_path = settings.DEFAULT_LANGUAGE
    if response.status_code == 404 and not language_from_path and i18n_patterns_used:
        language_path = '/%s%s' % (language, request.path_info)
        path_valid = is_valid_path(language_path, urlconf)
        path_needs_slash = (
            not path_valid and (
                settings.APPEND_SLASH and not language_path.endswith('/')
                and is_valid_path('%s/' % language_path, urlconf)
            )
        )

        if path_valid or path_needs_slash:
            script_prefix = get_script_prefix()
            # Insert language after the script prefix and before the
            # rest of the URL
            language_url = request.get_full_path(force_append_slash=path_needs_slash).replace(
                script_prefix,
                '%s%s/' % (script_prefix, language),
                1
            )
            return self.response_redirect_class(language_url)

    if not i18n_patterns_used and language_from_path:
        patch_vary_headers(response, ('Accept-Language',))
    if 'Content-Language' not in response:
        response['Content-Language'] = language
    return response
```

Do you think it makes sense?